// convoy.js/dist/convoy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhook = exports.Convoy = void 0;
var client_1 = require("./client");
var project_1 = require("./Api/project");
var subscription_1 = require("./Api/subscription");
var source_1 = require("./Api/source");
var event_delivery_1 = require("./Api/event-delivery");
var event_1 = require("./Api/event");
var endpoint_1 = require("./Api/endpoint");
var delivery_attempt_1 = require("./Api/delivery-attempt");
var webhook_1 = require("./webhook");
Object.defineProperty(exports, "Webhook", { enumerable: true, get: function () { return webhook_1.Webhook; } });
var sqs_1 = require("./Api/sqs");
var kafka_1 = require("./Api/kafka");
var Convoy = /** @class */ (function () {
    function Convoy(options) {
        this.client = new client_1.Client(options);
        this.projects = new project_1.Project(this.client);
        this.subscriptions = new subscription_1.Subscription(this.client);
        this.sources = new source_1.Source(this.client);
        this.events = new event_1.Event(this.client);
        this.eventDeliveries = new event_delivery_1.EventDelivery(this.client);
        this.endpoints = new endpoint_1.Endpoint(this.client);
        this.deliveryAttempts = new delivery_attempt_1.DeliveryAttempt(this.client);
        this.sqs = new sqs_1.SQS(this.client);
        this.kafka = new kafka_1.Kafka(this.client);
    }
    return Convoy;
}());
exports.Convoy = Convoy;
